File: Source/Core/Modes/InsertMode.cs
namespace Quip {
  class InsertMode : Mode {
    public InsertMode (DocumentView view)
      : base("Insert", view) {
      AddMapping(new Keystroke(Key.Escape), LeaveInsertMode);

      view.PushCursorStyle(CursorStyle.VerticalBar);
    }

    protected override bool OnHandleKey (Key input, DocumentView view) {
      if (input != Key.Backspace) {
        view.MoveTo(view.Document.Insert(input.ToText(), view.Cursor));
      } else {
        view.MoveTo(view.Document.Erase(view.Cursor));
      }

      return true;
    }

    bool LeaveInsertMode (DocumentView view) {
      view.PopCursorStyle();
      view.Mode = new NormalMode(view);
      return true;
    }
  }
}

File: Source/InsertMode.cpp
#include "InsertMode.hpp"

#include "Document.hpp"
#include "EditContext.hpp"
#include "KeyStroke.hpp"

namespace quip {
  InsertMode::InsertMode () {
    addMapping(Key::Escape, &InsertMode::commitInsert);
  }
  
  std::string InsertMode::status () const {
    return "Insert";
  }
  
  bool InsertMode::onUnmappedKey (const KeyStroke & keyStroke, EditContext & context) {
    if (keyStroke.text().size() > 0) {
      context.document().insert(context.selections(), keyStroke.text());
      return true;
    }
    
    return false;
  }
  
  void InsertMode::commitInsert (EditContext & context) {
    context.leaveMode();
  }
}